### PR TITLE
Make hashing independent of notebook content

### LINKED ISSUE #1712: different hashing in release_feedback and fetch_feedback
### Operating system
Ubuntu 20.04.5 LTS
### `nbgrader --version`
0.8.0
### `jupyterhub --version` (if used with JupyterHub)
2.3.1
### `jupyter notebook --version` 
6.4.12
### `jupyter lab --version`
3.4.6
### Expected behavior
same result from `notebook_hash` in `release_feedback.py` and `fetch_feedback.py`
### Actual behavior
different results, resulting in `[FetchFeedbackApp | WARNING] Could not find feedback for ...` during `nbgrader fetch_feedback --debug`
### Steps to reproduce the behavior
As instructor do `nbgrader release_feedback --debug <assignment>`. Then as student do `nbgradere fetch_feedback --debug <assignment>`.

## Explanation

We are using nbgrader with jupyterhub and jupyter lab for one of our courses much to our liking. For several months all was working well but very recently we stumbled on the following incident. Two students (of 20) are not able to fetch the feedback of one particular assignment.

After some debugging I found that the hash in `release_feedback.py` was generated with different values than in `fetch_feedback.py`:

**`nbgrader release_feedback --debug 'Assignment 3 Wavefront sensing`**

```raw
[ReleaseFeedbackApp | DEBUG] nbfile is: /home/grader-ap3122/ap3122/submitted/<username>/Assignment 3 Wavefront sensing/Assignment Wavefront aberrations.ipynb
[ReleaseFeedbackApp | DEBUG] Unique key is: ap3122+Assignment 3 Wavefront sensing+Assignment Wavefront aberrations+<username>+2022-11-11 15:17:36.574853 UTC
[ReleaseFeedbackApp | DEBUG] checksum is: f9990f2f5f029af5091cbc417c5a591e
[ReleaseFeedbackApp | DEBUG] dest is: /usr/local/share/nbgrader/exchange/ap3122/feedback/f9990f2f5f029af5091cbc417c5a591e.html
```

**`nbgrader fetch_feedback --debug 'Assignment 3 Wavefront sensing`**

```raw
[FetchFeedbackApp | DEBUG] notebook is: /data/jupyterhub/home/<username>/.local/share/jupyter/nbgrader_cache/ap3122/<username>+Assignment 3 Wavefront sensing+2022-11-11 15:17:36.574853 UTC/Assignment Wavefront aberrations.ipynb
[FetchFeedbackApp | DEBUG] Unique key is: ap3122+Assignment 3 Wavefront sensing+Assignment Wavefront aberrations+<username>+2022-11-11 15:17:36.574853 UTC
[FetchFeedbackApp | DEBUG] nb_hash is: 209d0cfb1a1a05ba8d64273f2fd04271
[FetchFeedbackApp | DEBUG] feedbackpath is: /usr/local/share/nbgrader/exchange/ap3122/feedback/209d0cfb1a1a05ba8d64273f2fd04271.html
```

*the student name has been replaced with `<username>`*

*I've added some debugging code to the two python-scripts logging the values of the appropriate variables*

Clearly the `nbfile` and `notebook` have very different value resulting in different hash values. This is reflected in the different values of `dest` and `feedbackpath` which should be the same.

### PR DESCRIPTION
**Pull request by tuncbkose**
Currently, changing the content of a submitted notebook breaks feedback distribution (see some discussion in https://github.com/jupyter/nbgrader/issues/1202). This is because feedback distribution uses a hash that depends on the notebook content.

This pr aims to solve this issue by duplicating the `timestamp.txt` mechanism. A `submission_secret.txt` file containing a random string is created during submission, and the notebook hash is changed to depend only on this file and notebook name.

Should fix #1712

### CODE REVIEW

**Comment by danielmaitre** [comment]
I like this solution and would also solve some of the issues I am facing!

File: nbgrader/exchange/default/fetch_feedback.py [hunk 1]
```diff
@@ -48,35 +48,57 @@ def init_src(self):
                     self.coursedir.course_id, assignment_id, timestamp))
             pattern = os.path.join(self.cache_path, submission, "*.ipynb")
             notebooks = glob.glob(pattern)
+
+            # Check if a secret is provided (new method)
```
Comment (brichet) [inline nbgrader/exchange/default/fetch_feedback.py:52]:
```suggestion
            # Check if a secret is provided
```

File: nbgrader/exchange/default/fetch_feedback.py [hunk 2]
```diff
@@ -48,35 +48,57 @@ def init_src(self):
...
-                nb_hash = notebook_hash(notebook)
-                feedbackpath = os.path.join(self.outbound_path, '{0}.html'.format(nb_hash))
-                if os.path.exists(feedbackpath):
-                    self.feedback_files.append((notebook_id, timestamp, feedbackpath))
-                    self.log.warning(
-                        "Found legacy feedback for '{}/{}/{}' submitted at {}".format(
-                            self.coursedir.course_id, assignment_id, notebook_id, timestamp))
-                    continue
+
+                # If a secret is provided (new method), use that
```
Comment (brichet) [inline nbgrader/exchange/default/fetch_feedback.py:62]:
```suggestion
                # If a secret is provided, use that secret
```

File: nbgrader/exchange/default/list.py [hunk 3]
```diff
@@ -128,23 +128,36 @@ def parse_assignments(self):
...
-                self.log.debug("Unique key is: {}".format(unique_key))
-                nb_hash = notebook_hash(notebook, unique_key)
-                exchange_feedback_path = os.path.join(
-                    self.root, info['course_id'], 'feedback', '{0}.html'.format(nb_hash))
-                has_exchange_feedback = os.path.isfile(exchange_feedback_path)
-                if not has_exchange_feedback:
-                    # Try looking for legacy feedback.
-                    nb_hash = notebook_hash(notebook)
+
+                # Check if a secret is provided (new method)
```
Comment (brichet) [inline nbgrader/exchange/default/list.py:132]:
```suggestion
                # Check if a secret is provided
```

**Review by brichet — COMMENTED** [review]
Thanks @tuncbkose, let's move forward and merge this one.
Can you rebase on main please ?

**Comment by tuncbkose** [comment]
Thanks for the review. I'll rebase within a day or two and ping you

**Comment by tuncbkose** [comment]
@brichet rebased.

**Comment by brichet** [comment]
Thanks @tuncbkose

**Comment by LukasMueller187** [comment]
Happy to see this. :tada:  Was manually applying the PR every update. Thanls @tuncbkose

**Pull request merged**